openedx/features/colaraz_features/middlewares.py

```diff
@@ -0,0 +1,84 @@
+from django.conf import settings
+from django.contrib.auth import get_user, logout
+from django.http import HttpResponseRedirect
+from django.urls import reverse
+
+import third_party_auth
+from third_party_auth import pipeline
+
+from student.helpers import get_next_url_for_login_page
+from student.models import UserAttribute
+
+
+class ColarazAuthenticationMiddleware(object):
+    """
+    Special middleware for Colaraz which does following tasks:
+    1) Verify User's accessability over his requested site
+    2) Login's User if he is logged-in on Colaraz's website
+    """
+
+    def _get_user_site(self, user, session):
+        """
+        Returns domain of site associated with the User.
+        """
+        if not session.get('user_site', None):
+            session['user_site'] = UserAttribute.get_user_attribute(user, 'created_on_site')
+        return session.get('user_site', None)
+
+    def _get_request_site(self, request):
+        """
+        Returns domain of site being requested by the User.
+        """
+        site = getattr(request, "site", None)
+        domain = getattr(site, "domain", None)
+        return domain
+
+    def _redirect_to_login(self, request):
+        """
+        Returns HttpResponseRedirect object, redirecting User to third-party-auth
+        identity provider on the basis of 'DEFAULT_AUTH_PROVIDER_BACKEND_NAME' or 
+        to 'login' page in-case auth provider isn't configured properly.
+        """
+        DEFAULT_AUTH_PROVIDER_BACKEND_NAME = getattr(settings, 'DEFAULT_AUTH_PROVIDER_BACKEND_NAME', None)
+
+        if third_party_auth.is_enabled() and DEFAULT_AUTH_PROVIDER_BACKEND_NAME:
+            provider = [enabled for enabled in third_party_auth.provider.Registry.enabled()
```

Comment:
Why we are iterating the list here? Can't we get the only desired provider? like get_queryset?

Reply:
There are multiple authentication types like OAUTH and SAML etc. The registry we are seeing here is not a model, its kind of a handler which searches for enabled IdP from OAuth2ProviderConfig and SAMLConfiguration etc.
If we directly take our configuration from OAuth2ProviderConfig it will work absolutely fine, but it will ignore other providers (SAMLConfiguration and LTIProviderConfig), that's why to make it more generic I used the builtin filtering.